### PR TITLE
Get rid of instanceof in StringOrArray extension

### PR DESCRIPTION
**Pull request by herndlm**

### CODE REVIEW

**Comment by IanDelMar** [comment]
I'm not quite sure that this is sufficient. While `$keyType instanceof StringType` was a boolean `$keyType->isString()` is a trinary logic. We should cover `$keyType->isString()->maybe()` as well, don't we?

**Comment by herndlm** [comment]
Not sure tbh, this does not change behaviour at least to before, except e.g. supporting unions of strings too. Maybes like mixed were also not considered before. It could still be improved any time I guess

File: src/StringOrArrayDynamicFunctionReturnTypeExtension.php [hunk 1]
```diff
@@ -35,7 +35,7 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
         $dataArgType = $scope->getType($dataArg);
         if ($dataArgType->isArray()->yes()) {
             $keyType = $dataArgType->getIterableKeyType();
-            if ($keyType instanceof StringType) {
+            if ($keyType->isString()->yes()) {
                 return new ArrayType(new StringType(), new StringType());
             }
             return new ArrayType(new IntegerType(), new StringType());
```
Comment (IanDelMar) [inline src/StringOrArrayDynamicFunctionReturnTypeExtension.php:41]:
```suggestion
            $returnType = [new ArrayType(new IntegerType(), new StringType())];
            if ($keyType->isString()->maybe()) {
                $returnType[] = new ArrayType(new StringType(), new StringType());
            }
            return TypeCombinator::union(...$returnType);
```

**Review by IanDelMar — CHANGES_REQUESTED** [review]

File: src/StringOrArrayDynamicFunctionReturnTypeExtension.php (hunk 1 above)
Comment (IanDelMar) [inline src/StringOrArrayDynamicFunctionReturnTypeExtension.php:41]:
```suggestion
            if ($keyType->isString()->no()) {
                return new ArrayType(new IntegerType(), new StringType());
            }
            return TypeCombinator::union(
                new ArrayType(new StringType(), new StringType()),
                new ArrayType(new IntegerType(), new StringType())
            );
```

**Review by IanDelMar — CHANGES_REQUESTED** [review]

**Comment by IanDelMar** [comment]
I agree, it is as good as before, it just makes it more obvious that the unknown key type case was not covered.

File: src/StringOrArrayDynamicFunctionReturnTypeExtension.php (hunk 1 above)
Comment (herndlm) [reply src/StringOrArrayDynamicFunctionReturnTypeExtension.php:41]: I see what you're going for I think. I don't really understand though what this extension is even doing :) I was just replacing stuff without changing much behaviour.
Why does it not just pass through the array but recreate it for example. Unfortunately not  on my computer anymore to dig deeper..

Comment (IanDelMar) [reply src/StringOrArrayDynamicFunctionReturnTypeExtension.php:41]: It sets the value type to string no matter what. However, this is not what the extension was intended to I think. See #154.

Comment (herndlm) [reply src/StringOrArrayDynamicFunctionReturnTypeExtension.php:41]: I'll check it later and try to add tests too it not existing yet. But does it make sense to you to directly pass through the key type? It should noy be modified, right?

**Pull request closed without merging**